### DISCUSSION TITLE
Add a py.typed

### DISCUSSION THREAD

**Discussion by sthiele** [potassco/python-project-template#14] [Ideas]
For `clintest` there was recently the [request](https://github.com/potassco/clintest/issues/22) to add a `py.typed` file to enable type checking in other packages. I wasn't aware that this is necessary.
Should this also be part of the template?

**Comment by MaxOstrowski** [comment]
Very interesting. Is 
`include_package_data = True`
and
```
[options.packages.find]
where = src
```
enough to include this in the installation process ?

**Reply by sthiele** [reply]
Sorry, I have no clue. But I according to this https://stackoverflow.com/questions/76073605/add-py-typed-as-package-data-with-setuptools-in-pyproject-toml

This should work 
```
[tool.setuptools.package-data]
"pkgname" = ["py.typed"]

[tool.setuptools.packages.find]
where = ["src"]
```

**Reply by MaxOstrowski** [reply]
According to:
https://setuptools.pypa.io/en/latest/userguide/datafiles.html#include-package-data
it should automatically be included as package data because we seem to use
https://pypi.org/project/setuptools-scm/
as part of the build tools, which apparently should add all files that are managed by git to the build.
That doesn't sound too compelling to me, as the build process will be dependent on the .git files, but apparently this is a thing...

Someone with experience and an opinion here ?

**Reply by sthiele** [reply]
Packaging in Python sucks.

**Reply by MaxOstrowski** [reply]
Thats an opinion and I agree.
I haven't found any criticism on `setuptools-scm` and I still think its odd to rely on the revisioning tool for building the package.
@susuhahnml @rkaminsk any opinion about it ?

**Reply by rkaminsk** [reply]
I think it's fine. The source tarball does not need git. We only need git to create the source tarball.

**Reply by rkaminsk** [reply]
I'll go over the PRs next and make sure the py.typed thing is added. It probably takes less time to fix than discussing about it.

**Reply by susuhahnml** [reply]
What do you mean by "rely on the revisioning tool for building the package" sorry, I am not following :(

**Reply by rkaminsk** [reply]
> Packaging in Python sucks.

Packaging pure Python packages is actually not that complicated. Of course one has to take care of all the optional things one wants to add. It only gets (really) messy once you want to ship projects that need a compiler. However, there has been quite some development to even make this possible these days.

**Comment by sthiele** [comment]
@susuhahnml because
> [setuptools-scm](https://github.com/pypa/setuptools_scm) extracts Python package versions from git or hg metadata instead of declaring them as the version argument or in an SCM managed file.

This means if we switch to [another versioning tool](https://pijul.org/), then packaging also needs to be adjusted.
My instincts tell me to reduce unnecessary dependencies ... but this seems to be a thing :shrug:

**Reply by rkaminsk** [reply]
At the moment, we do not use it to extract the version - just to get the list of files for the source tarball.

In my experience, this does exactly what is required by just adding a tiny package. However, there are other ways. We could for example use manifest files. In my experience, this easily leads to broken source tarballs.

I would like to keep it as is.
